Core/Src/Systems/External/BMS.c

```diff
@@ -0,0 +1,118 @@
+#include "../../../Inc/Systems/External/BMS.h"
+
+#include <stdio.h>  // For printf
+#include "../../../Inc/Systems/External/Can/Can.h"
+
+void initBms(Bms* bms, int hz, const char* dbcFn) {
+    if (bms == NULL) return;
+
+    initExternalSystem(&bms->extSystem, "Bms", hz, 0);
+
+    bms->packVoltage = 0.0f;
+    bms->packCurrent = 0.0f;
+    bms->stateOfCharge = 0.0f;
+    bms->cellVoltageMin = 0.0f;
+    bms->cellVoltageMax = 0.0f;
+    bms->cellTemperatureMin = 0.0f;
+    bms->cellTemperatureMax = 0.0f;
+    bms->totalPackCapacity = 0.0f;
+    bms->remainingPackCapacity = 0.0f;
+    bms->packHealth = 0.0f;
+    bms->chargeStatus = IDLE;
+    bms->extSystem.system.updateable.update = updateBms;
+
+    DBC dbc;
+    if(!parseDbcFile(&dbc, dbcFn)) {
+        printf("Error: Couldn't parse dbc.\n");
+        return;
+    }
+    bms->dbc = &dbc;
+
+    Message* messageMap[MAX_MESSAGES] = {NULL};
+
+    int size = dbc.messageCount;
+    for (int i = 0; i < size; i++) {
+        int index = dbc.messages[i].id % MAX_MESSAGES;
+        messageMap[index] = &dbc.messages[i];
```

Comment:
Using modulo for index calculation means different message IDs could map to the same index (e.g., if ```MAX_MESSAGES``` is 100, IDs 105 and 205 would both map to index 5). 
If the IDs are designed in a way that would prevent same modulo values then it is fine, but otherwise there is a potential collision risk in message map indexing.

Potential fixes include: 
1. Adding collision detection like:
```c
if (messageMap[index] != NULL) {
    // Handle collision - either error out or implement proper handling
    printf("Warning: Message ID collision at index %d\n", index);
}
```
2. Or use a more robust mapping structure like a hash table with collision resolution.